specs/signing-and-verification-workflow.md

```diff
@@ -70,3 +70,46 @@ The user wants to pull an OCI artifact only if they are signed by a trusted publ
+    1. Construct the blob payload as defined in [`signature specification`](./signature-specification.md#payload)
+    1. Verify that the signing certificate is valid and satisfies [certificate requirements](./signature-specification.md#certificate-requirements).
+    1. Verify that the signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).
+    1. Generate signature.
+        1. Generate signature using a signature format specified in [supported signature envelopes](./signature-specification.md#supported-signature-envelopes). Also, as part of this step, the user-defined/supplied custom attributes should be added to the annotations of the signature payload.
+        1. If the user wants to timestamp the signature, obtain an [RFC-3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant timestamp for the signature generated in the previous step. Otherwise, continue to the next step.
+            1. Verify that the timestamp signing certificate satisfies [certificate requirements](./signature-specification.md#certificate-requirements).
+            1. Verify that the timestamp signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).
+        1. Embed timestamp to the signature envelope.
+1. **Save the signature envelope:** Save the signature envelope generated in the previous step to a file. File extension should be the original blob file name plus `.sig.jws` for JWS signatures and `.sig.cose` for COSE signatures.
+
+## Arbitrary blob verification workflow
+
+The user wants to consume an arbitrary blob only if it was signed by a trusted publisher and the signature associated with the blob is valid.
+
+### Verification Prerequisites
+
+- User has the blob that they want to consume, along with its detached signature.
+- User has configured [trust store and trust policy](./trust-store-trust-policy.md) required for signature verification.
+
+### Verification Steps
+
+1. **Should implementations of this specification verify the signature? :** Depending upon [trust-policy](./trust-store-trust-policy.md#blob-trust-policy) configuration, determine whether implementations of this specification need to verify the signature or not.
+   If signature verification should be skipped for the given blob, skip the below steps and directly jump to step 4.
+1. **Verify the detached signature:**
+    1. Parse and validate the signature envelope using the detached signature's file extension as the envelope type.
+    1. Verify the signature envelope using trust-store and trust-policy as mentioned in [signature evaluation](./trust-store-trust-policy.md#signature-evaluation) section.
+    1. If the signature verification fails, exit.
+1. Construct the blob payload as defined in [`signature specification`](./signature-specification.md#payload)
```

Comment:
Since the payload is a JSON object, how do you make sure that the newly constructed blob payload is exactly the same as the payload constructed by the signer?

How to canonicalize the JSON object? For instance, `{"foo":1,"bar":2}` and `{"bar":2,"foo":1}` are identical JSON objects but different strings.

How do verifiers deal with user metadata?

Reply:
It does not seem right. Should this step be something like "extract the blob payload from the signature envelope"?

Reply:
> Since the payload is a JSON object, how do you make sure that the newly constructed blob payload is exactly the same as the payload constructed by the signer?

> How to canonicalize the JSON object? For instance, {"foo":1,"bar":2} and {"bar":2,"foo":1} are identical JSON objects but different strings.

Notation is not verifying the string equality of the original payload vs the signed payload (in Json). The workflow here is calling out the verification on Size, MediaType and the digest. Which is what is being done for OCI artifacts as well. See https://github.com/oras-project/oras-go/blob/be5973602aa86b007b948be27ddc9c5749580b34/content/descriptor.go#L38

> It does not seem right. Should this step be something like "extract the blob payload from the signature envelope"?

Take a look at [this code](https://github.com/notaryproject/notation-go/blob/main/verifier/verifier.go#L173). In the case of OCI, original payload is already available (from the registry) and we are comparing it against the payload retrieved from the signature envelope. However, that is not the case for blobs. Hence, we need to construct the original payload from the blob itself and then compare it against `targetArtifact.payload`.

> How do verifiers deal with user metadata?

Good call out. Will add it.

Reply:
I modified the language a little bit to avoid any confusion. Let me know if you still have any concerns.